src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -343,7 +361,9 @@ impl Qualif for HasMutInterior {
     }
 }
 
-// Constant containing an ADT that implements Drop.
+/// Constant containing an ADT that implements Drop.
+/// This must be ruled out because we cannot run `Drop` during compile-time
+/// as that might not be a `const fn`.
```

Comment:
And because promoting `&SomethingThatNeedsDrop` would remove side-effects of dropping that value.
Again, the point about not being able to observe behavioral differences after promotion.